modules/kafka/kafka-dest-worker.c

```diff
@@ -56,21 +58,63 @@ _format_message_and_key(KafkaDestWorker *self, LogMessage *msg)
                         self->super.seq_num, NULL, self->key);
 }
 
+static rd_kafka_topic_t *
+_calculate_topic_from_template(KafkaDestWorker *self, LogMessage *msg)
+{
+  KafkaDestDriver *owner = (KafkaDestDriver *) self->super.owner;
+  log_template_format(owner->topic_name, msg, &owner->template_options, LTZ_SEND, self->super.seq_num, NULL,
+                      self->topic_name_buffer);
+  rd_kafka_topic_t *topic = kafka_dd_query_insert_topic(owner, self->topic_name_buffer->str);
+
+  if (!topic)
+    {
+      /* if topic name is invalid, then topic is fallback topic which we know is defined*/
+      topic = owner->fallback_topic;
+      g_string_assign(self->topic_name_buffer, owner->fallback_topic_name);
+    }
+
+  return topic;
+}
+
+static rd_kafka_topic_t *
+_get_literal_topic(KafkaDestWorker *self, LogMessage *msg)
+{
+  KafkaDestDriver *owner = (KafkaDestDriver *) self->super.owner;
+
+  g_string_assign(self->topic_name_buffer, owner->topic_name->template);
+  return owner->topic;
+}
+
+static rd_kafka_topic_t *
+_calculate_topic(KafkaDestWorker *self, LogMessage *msg)
+{
+  KafkaDestDriver *owner = (KafkaDestDriver *) self->super.owner;
+
+  if (kafka_dd_is_topic_name_a_template(owner))
+    {
+      return _calculate_topic_from_template(self, msg);
+    }
+
+  return _get_literal_topic(self, msg);
+}
+
+
 static gboolean
 _publish_message(KafkaDestWorker *self, LogMessage *msg)
 {
   KafkaDestDriver *owner = (KafkaDestDriver *) self->super.owner;
   int block_flag = _is_poller_thread(self) ? 0 : RD_KAFKA_MSG_F_BLOCK;
+  rd_kafka_topic_t *topic = _calculate_topic(self, msg);
 
-  if (rd_kafka_produce(owner->topic,
+  if (rd_kafka_produce(topic,
                        RD_KAFKA_PARTITION_UA,
                        RD_KAFKA_MSG_F_FREE | block_flag,
                        self->message->str, self->message->len,
                        self->key->len ? self->key->str : NULL, self->key->len,
                        log_msg_ref(msg)) == -1)
     {
       msg_error("kafka: failed to publish message",
-                evt_tag_str("topic", owner->topic_name),
+                evt_tag_str("topic", self->topic_name_buffer->str),
```

Comment:
There is a getter function in `rdkafka`, which returns the `topic_name` of the `rd_kafka_topic_t` object.
That can handle the templated topic, the literal topic and the fallback topic, too, in just one function call.

https://docs.confluent.io/2.0.0/clients/librdkafka/rdkafka_8h.html#a3ea8248521e63b4afb1d292fbcc584e2